core/tee/tee_fs_common.c

```diff
@@ -37,19 +40,939 @@
+{
+	char block_path[REE_FS_NAME_MAX];
+
+	get_block_filepath(fdp->filename, block_num, 0, block_path);
+	DMSG("%s", block_path);
+
+	return ree_fs_open(block_path, TEE_FS_O_CREATE | TEE_FS_O_WRONLY);
+}
+
+static int free_block(struct tee_fs_fd *fdp, int block_num)
+{
+	char block_path[REE_FS_NAME_MAX];
+	uint8_t version =
+		get_backup_version_of_block(fdp->meta, block_num);
+
+	get_block_filepath(fdp->filename, block_num, version, block_path);
+	DMSG("%s", block_path);
+
+	return ree_fs_unlink(block_path);
+}
+
+/*
+ * encrypted_fek: as input for META_FILE and BLOCK_FILE
+ */
+static int encrypt_and_write_file(int fd,
+		enum tee_fs_file_type file_type,
+		void *data_in, size_t data_in_size,
+		uint8_t *encrypted_fek)
+{
+	TEE_Result tee_res;
+	int res, bytes;
+	uint8_t *ciphertext;
+	size_t header_size = tee_fs_get_header_size(file_type);
+	size_t ciphertext_size = header_size + data_in_size;
+
+	ciphertext = malloc(ciphertext_size);
+	if (!ciphertext)
+		return -1;
+
+	tee_res = tee_fs_encrypt_file(file_type,
+			data_in, data_in_size,
+			ciphertext, &ciphertext_size, encrypted_fek);
+	if (tee_res != TEE_SUCCESS) {
+		EMSG("error code=%x", tee_res);
+		res = -1;
+		goto fail;
+	}
+
+	bytes = ree_fs_write(fd, ciphertext, ciphertext_size);
+	if (bytes != (int)ciphertext_size) {
+		EMSG("bytes(%d) != ciphertext size(%zu)",
+				bytes, ciphertext_size);
+		res = -1;
+		goto fail;
+	}
+
+	res = bytes - header_size;
+
+fail:
+	free(ciphertext);
+
+	return res;
+}
+
+/*
+ * encrypted_fek: as output for META_FILE
+ *                as input for BLOCK_FILE
+ */
+static int read_and_decrypt_file(int fd,
+		enum tee_fs_file_type file_type,
+		void *data_out, size_t *data_out_size,
+		uint8_t *encrypted_fek)
+{
+	TEE_Result tee_res;
+	int res = 0;
+	int bytes;
+	void *ciphertext = NULL;
+	size_t file_size;
+	size_t header_size = tee_fs_get_header_size(file_type);
+
+	res = get_file_length(fd, &file_size);
+	if (res < 0)
+		return res;
+
+	TEE_ASSERT(file_size >= header_size);
+
+	ciphertext = malloc(file_size);
+	if (!ciphertext)
+		return -1;
+
+	bytes = ree_fs_read(fd, ciphertext, file_size);
+	if (bytes != (int)file_size) {
+		res = -1;
+		goto fail;
+	}
+
+	tee_res = tee_fs_decrypt_file(file_type,
+			ciphertext, file_size,
+			data_out, data_out_size,
+			encrypted_fek);
+	if (tee_res != TEE_SUCCESS)
+		res = -1;
+
+fail:
+	free(ciphertext);
+
+	return res;
+}
+
+static int create_meta_file(const char *filename,
+		struct tee_fs_file_meta *new_meta, int version)
+{
+	int res, fd = -1;
+	char meta_path[REE_FS_NAME_MAX];
+
+	get_meta_filepath(filename, version, meta_path);
+
+	fd = ree_fs_open(meta_path, TEE_FS_O_CREATE | TEE_FS_O_WRONLY);
+	if (fd < 0)
+		return -1;
+
+	res = encrypt_and_write_file(fd, META_FILE,
+			(void *)&new_meta->info, sizeof(new_meta->info),
+			new_meta->encrypted_fek);
+
+	ree_fs_close(fd);
+	return res;
+}
+
+static struct tee_fs_file_meta *tee_fs_file_create(const char *file,
```

Comment:
Since this function only handles meta-files I would rename this function to `tee_fs_meta_create`. By doing so it would also be more similar to `tee_fs_meta_open` etc...